File: plugins/antigravity-awesome-skills/skills/whatsapp-cloud-api/assets/boilerplate/nodejs/src/webhook-handler.ts
import crypto from 'crypto';
import { Request, Response, NextFunction } from 'express';
import { WebhookPayload, IncomingMessage, StatusUpdate } from './types';

const SAFE_CHALLENGE_RE = /^[A-Za-z0-9._-]{1,200}$/;

/**
 * Middleware para validar assinatura HMAC-SHA256 dos webhooks do WhatsApp.
 *
 * A Meta assina cada request com o App Secret. Validar esta assinatura
 * previne que requests falsificados sejam processados. Usar comparacao
 * constant-time previne timing attacks.
 *
 * IMPORTANTE: Este middleware deve ser aplicado ANTES de qualquer body parser
 * que altere o raw body (express.json() por exemplo).
 */
export function validateHMAC(appSecret: string) {
  return (req: Request, res: Response, next: NextFunction): void => {
    const signature = req.headers['x-hub-signature-256'] as string;

    if (!signature) {
      console.warn('Webhook request without signature header');
      res.sendStatus(401);
      return;
    }

    const rawBody = (req as any).rawBody;
    if (!rawBody) {
      console.error('Raw body not available. Ensure rawBody middleware is configured.');
      res.sendStatus(500);
      return;
    }

    const expectedSignature =
      'sha256=' +
      crypto.createHmac('sha256', appSecret).update(rawBody).digest('hex');

    const isValid = crypto.timingSafeEqual(
      Buffer.from(signature),
      Buffer.from(expectedSignature)
    );

    if (!isValid) {
      console.warn('Invalid webhook signature');
      res.sendStatus(401);
      return;
    }

    next();
  };
}

/**
 * Middleware para capturar o raw body antes do JSON parse.
 * Necessario para validacao HMAC.
 */
export function rawBodyMiddleware(req: Request, _res: Response, buf: Buffer): void {
  (req as any).rawBody = buf;
}

/**
 * Handler de verificacao do webhook (GET).
 * A Meta envia um challenge que deve ser retornado para confirmar o endpoint.
 */
export function handleWebhookVerification(verifyToken: string) {
  return (req: Request, res: Response): void => {
    const mode = req.query['hub.mode'] as string;
    const token = req.query['hub.verify_token'] as string;
    const challenge = req.query['hub.challenge'] as string;

    if (mode === 'subscribe' && token === verifyToken && SAFE_CHALLENGE_RE.test(challenge)) {
      console.log('Webhook verified successfully');
      res.type('text/plain').status(200).send(challenge);
    } else {
      console.warn('Webhook verification failed: invalid token');
      res.status(mode === 'subscribe' && token === verifyToken ? 400 : 403).send();
    }
  };
}

/**
 * Extrai mensagens e status updates do payload do webhook.
 */
export function parseWebhookPayload(payload: WebhookPayload): {
  messages: IncomingMessage[];
  statuses: StatusUpdate[];
} {
  const messages: IncomingMessage[] = [];
  const statuses: StatusUpdate[] = [];

  for (const entry of payload.entry || []) {
    for (const change of entry.changes || []) {
      if (change.value.messages) {
        messages.push(...change.value.messages);
      }
      if (change.value.statuses) {
        statuses.push(...change.value.statuses);
      }
    }
  }

  return { messages, statuses };
}

/**
 * Extrai o texto ou ID de selecao de uma mensagem recebida.
 */
export function extractMessageContent(message: IncomingMessage): {
  type: string;
  text?: string;
  buttonId?: string;
  listId?: string;
  mediaId?: string;
  location?: { latitude: number; longitude: number };
  reaction?: { emoji: string; messageId: string };
} {
  switch (message.type) {
    case 'text':
      return { type: 'text', text: message.text?.body };

    case 'interactive':
      if (message.interactive?.type === 'button_reply') {
        return {
          type: 'button',
          buttonId: message.interactive.button_reply?.id,
          text: message.interactive.button_reply?.title,
        };
      }
      if (message.interactive?.type === 'list_reply') {
        return {
          type: 'list',
          listId: message.interactive.list_reply?.id,
          text: message.interactive.list_reply?.title,
        };
      }
      if (message.interactive?.type === 'nfm_reply') {
        return {
          type: 'flow',
          text: message.interactive.nfm_reply?.response_json,
        };
      }
      return { type: 'interactive' };

    case 'image':
      return { type: 'image', mediaId: message.image?.id };

    case 'document':
      return { type: 'document', mediaId: message.document?.id };

    case 'video':
      return { type: 'video', mediaId: message.video?.id };

    case 'audio':
      return { type: 'audio', mediaId: message.audio?.id };

    case 'location':
      return {
        type: 'location',
        location: message.location
          ? { latitude: message.location.latitude, longitude: message.location.longitude }
          : undefined,
      };

    case 'reaction':
      return {
        type: 'reaction',
        reaction: message.reaction
          ? { emoji: message.reaction.emoji, messageId: message.reaction.message_id }
          : undefined,
      };

    default:
      return { type: message.type };
  }
}

File: plugins/antigravity-awesome-skills-claude/skills/whatsapp-cloud-api/assets/boilerplate/python/webhook_handler.py
"""WhatsApp Webhook Handler with HMAC-SHA256 validation."""

import hashlib
import hmac
import os
import re
from functools import wraps
from typing import Any

from flask import Response, abort, request


_SAFE_CHALLENGE_RE = re.compile(r"^[A-Za-z0-9._-]{1,200}$")


def validate_hmac_signature(app_secret: str | None = None):
    """
    Flask decorator to validate HMAC-SHA256 signature on webhook requests.

    A Meta assina cada request com o App Secret no header X-Hub-Signature-256.
    Validar esta assinatura previne requests falsificados.
    Usa hmac.compare_digest para comparacao constant-time (previne timing attacks).
    """
    secret = app_secret or os.environ["APP_SECRET"]

    def decorator(f):
        @wraps(f)
        def decorated_function(*args, **kwargs):
            signature = request.headers.get("X-Hub-Signature-256", "")
            if not signature:
                abort(401, "Missing signature header")

            raw_body = request.get_data()
            expected = "sha256=" + hmac.new(
                secret.encode(), raw_body, hashlib.sha256
            ).hexdigest()

            if not hmac.compare_digest(signature, expected):
                abort(401, "Invalid signature")

            return f(*args, **kwargs)

        return decorated_function

    return decorator


def verify_webhook(verify_token: str | None = None):
    """
    Handle webhook verification (GET request from Meta).
    Returns the challenge to confirm the webhook endpoint.
    """
    token = verify_token or os.environ["VERIFY_TOKEN"]

    mode = request.args.get("hub.mode")
    req_token = request.args.get("hub.verify_token")
    challenge = request.args.get("hub.challenge")

    if mode != "subscribe" or req_token != token:
        abort(403, "Verification failed")

    if not challenge or not _SAFE_CHALLENGE_RE.fullmatch(challenge):
        abort(400, "Invalid challenge")

    return Response(challenge, status=200, mimetype="text/plain")


def parse_webhook_payload(data: dict[str, Any]) -> dict[str, list]:
    """
    Parse webhook payload and extract messages and status updates.

    Returns dict with 'messages' and 'statuses' lists.
    """
    messages: list[dict] = []
    statuses: list[dict] = []

    for entry in data.get("entry", []):
        for change in entry.get("changes", []):
            value = change.get("value", {})
            if "messages" in value:
                messages.extend(value["messages"])
            if "statuses" in value:
                statuses.extend(value["statuses"])

    return {"messages": messages, "statuses": statuses}


def extract_message_content(message: dict[str, Any]) -> dict[str, Any]:
    """
    Extract readable content from an incoming message.

    Returns dict with 'type' and relevant fields (text, button_id, media_id, etc.).
    """
    msg_type = message.get("type", "unknown")

    if msg_type == "text":
        return {"type": "text", "text": message.get("text", {}).get("body")}

    if msg_type == "interactive":
        interactive = message.get("interactive", {})
        int_type = interactive.get("type")

        if int_type == "button_reply":
            reply = interactive.get("button_reply", {})
            return {"type": "button", "button_id": reply.get("id"), "text": reply.get("title")}

        if int_type == "list_reply":
            reply = interactive.get("list_reply", {})
            return {"type": "list", "list_id": reply.get("id"), "text": reply.get("title")}

        if int_type == "nfm_reply":
            return {
                "type": "flow",
                "text": interactive.get("nfm_reply", {}).get("response_json"),
            }

        return {"type": "interactive"}

    if msg_type in ("image", "document", "video", "audio"):
        media_data = message.get(msg_type, {})
        return {"type": msg_type, "media_id": media_data.get("id")}

    if msg_type == "location":
        loc = message.get("location", {})
        return {
            "type": "location",
            "latitude": loc.get("latitude"),
            "longitude": loc.get("longitude"),
        }

    if msg_type == "reaction":
        reaction = message.get("reaction", {})
        return {
            "type": "reaction",
            "emoji": reaction.get("emoji"),
            "message_id": reaction.get("message_id"),
        }

    return {"type": msg_type}
